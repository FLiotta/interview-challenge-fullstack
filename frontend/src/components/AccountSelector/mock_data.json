[
  {
    "id": 1,
    "name": "Vacaciones",
    "currency": {
      "id": 1,
      "name": "Dolar",
      "symbol": "$"
    },
    "funds": 2500
  },
  {
    "id": 2,
    "name": "Caja de ahorro",
    "currency": {
      "id": 1,
      "name": "Dolar",
      "symbol": "$"
    },
    "funds": 500
  },
  {
    "id": 3,
    "name": "Freelance EU",
    "currency": {
      "id": 2,
      "name": "Euro",
      "symbol": "â‚¬"
    },
    "funds": 932
  }
]