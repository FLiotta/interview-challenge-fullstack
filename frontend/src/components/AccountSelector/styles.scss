@import '../../styles/colors.scss';

.account-selector {
  width: 500px;
  height: 150px;
  border: 1px solid #eee;
  border-radius: 8px;
  display: flex;
  align-items: center;
  padding: 25px;
  cursor: pointer;
  background-color: #fff;
  transition: .5s all;
  position: relative;

  &:hover {
    background-color: #fbfbfb;
    box-shadow:
      0 3.5px 5.3px rgba(0, 0, 0, 0.028),
      0 11.8px 17.9px rgba(0, 0, 0, 0.042),
      0 53px 80px rgba(0, 0, 0, 0.07)
    ;
  }

  &__image {
    height: 100px;
    width: 100px;
    border: 2px dotted $brand_color;
    border-radius: 50%;
    display: flex;
    justify-content: center;
    align-items: center;

    &-symbol {
      margin-bottom: 0px;
      font-family: 'montserrat';
      font-weight: 900;
    }
  }

  &__info {
    flex-grow: 1;
    padding: 0 15px;
    height: 100%;
    display: flex;
    flex-direction: column;
    justify-content: space-between;
  }

  &__name {
    margin-bottom: 0px;
    font-family: 'montserrat';
    font-weight: 700;
  }

  &__date {
    margin-bottom: 0px;
    color: #bfbfbf;
  }

  &__funds {
    margin-bottom: 0px;
    line-height: 1;
  }

  &__placeholder {
    flex-grow: 1;
    padding: 0 15px;
    height: 100%;
    display: flex;
    align-items: center;

    &-text {
      margin-bottom: 0px;
      color: $brand_color;
      font-family: 'montserrat';
    }
  }
}

.account-dropdown {
  width: 100%;
  height: 250px;
  overflow-y: auto;
  background-color: #fff;
  position: absolute;
  left: 0;
  top: 160px;
  border: 1px solid $brand_color;
  z-index: 10;
  border-radius: 8px;
  box-shadow:
    0 3.5px 5.3px rgba(0, 0, 0, 0.028),
    0 11.8px 17.9px rgba(0, 0, 0, 0.042),
    0 53px 80px rgba(0, 0, 0, 0.07)
  ;

  &__item {
    display: flex;
    padding: 25px;
    display: flex;
    align-items: center;
    transition: 0.3s all;

    &:hover {
      background-color: #eee;
    }

    &--selected {
      background-color: $brand_color;
      color: #fff;

      .account-dropdown__item-symbol {
        background-color: $brand_color_light;
        color: $brand_color;
      }

      &:hover {
        background-color: $brand_color;
      }
    }

    &-name {
      margin-bottom: 0px;
      line-height: 1;
      margin-left: 15px;
    }

    &-funds {
      margin-bottom: 0px;
      line-height: 1;
      margin-left: auto;
    }

    &-symbol {
      padding: 7px 11px;
      background-color: $brand_color;
      border-radius: 50%;
      margin-bottom: 0px;
      line-height: 1;
      color: #fff;
    }
  }
}