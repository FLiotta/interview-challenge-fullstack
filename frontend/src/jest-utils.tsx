// @Packages
import { ReactNode } from 'react';
import { render as rtlRender } from '@testing-library/react';
import { Provider } from 'react-redux';

// @Project
import store from 'store';
import { JsxElement } from 'typescript';

function render(
  ui: ReactNode | JsxElement | any,
  renderOptions: any = { }
) {
  function Wrapper({ children }: any) {
    return (
      <Provider store={store}>{children}</Provider>
    )
  }
  return rtlRender(ui, {
     wrapper: Wrapper, 
     ...renderOptions
    })
}

export * from '@testing-library/react'

export { render }
