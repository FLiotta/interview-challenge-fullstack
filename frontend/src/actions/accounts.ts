// @Packages
import { AxiosResponse } from 'axios';

// @Project
import AccountService from 'services/account';
import { BackendResponse, IThunkDispatch, Account } from 'interfaces';


export const ACCOUNTS_FETCH = '[ACCOUNTS] FETCH';
export const ACCOUNTS_UPDATE_PRICE = '[ACCOUNTS] UPDATE PRICE';
export const ACCOUNTS_APPEND = '[ACCOUNTS] APPEND ACCOUNTS';

export const fetchAccounts = () => {
  return (dispatch: IThunkDispatch) => {

    return AccountService.getAllAccounts()
      .then((resp: AxiosResponse<BackendResponse<Account[]>>) => {
        const { metadata } = resp.data
        const { data } = resp.data;

        dispatch({
          type: ACCOUNTS_FETCH,
          payload: {
            data: data,
            total: metadata?.total
          }
        });
      })
  }
}

export const appendAccounts = (accs: Account[]) => ({
  type: ACCOUNTS_APPEND,
  payload: accs
});

export const updateAccountValue = (
  accountId: number, 
  amount: number, 
  operation: 'increase' | 'decrease'
) => ({
  type: ACCOUNTS_UPDATE_PRICE,
  payload: {
    accountId,
    amount,
    operation
  }
});