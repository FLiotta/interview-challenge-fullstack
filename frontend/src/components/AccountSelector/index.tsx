// @Packages
import React, { useEffect, useRef, useState } from 'react';
import { useSelector } from 'react-redux';
import cn from 'classnames';
import ReactTooltip from 'react-tooltip';
import dayjs from 'dayjs';

// @Project
import { useClickOutside } from 'hooks/useClickOutside';
import { Account } from 'interfaces';
import { selectAccounts } from 'selectors/accounts';

// @Own
import MockData from './mock_data.json';
import './styles.scss';

interface IAccountSelectorDropdown {
  selectedAccountId?: number,
  onAccountSelect: (acc: Account) => void
}

interface IAccountSelector {
  onAccountSelect: (acc: Account) => void
}

const AccountSelectorList: React.FC<IAccountSelectorDropdown> = ({
  selectedAccountId,
  onAccountSelect
}) => {
  const accounts: Account[] = useSelector(selectAccounts);

  return (
    <div className="account-dropdown">
      {accounts.map((acc) => (
        <div
          key={`${acc.id}`}
          onClick={() => onAccountSelect(acc)}
          className={cn("account-dropdown__item", {
            "account-dropdown__item--selected": acc.id === selectedAccountId
          })}
        >
          <div className="account-dropdown__item-symbol">
            <span>{acc.currency?.symbol}</span>
          </div>
          <h5 className="account-dropdown__item-name">{acc.name}</h5>
          <h5 className="account-dropdown__item-funds">${acc.funds}</h5>
        </div>
      ))}
    </div>
  )
}

const AccountSelector: React.FC<IAccountSelector> = ({
  onAccountSelect
}) => {  
  const [selectedAccount, setSelectedAccount] = useState<Account | undefined>();
  const [dropdownVisible, setDropdownVisible] = useState<boolean>(false);
  const AccountSelectorRef = useRef<any>();

  useClickOutside(AccountSelectorRef, () => {
    setDropdownVisible(false)
  });

  const handleOnAccountSelect = (acc: Account) => {
    setDropdownVisible(false);
    onAccountSelect(acc);
    setSelectedAccount(acc);
  }

  useEffect(() => {
    ReactTooltip.rebuild();
  }, [selectedAccount])

  return (
    <div
      className="account-selector"
      onClick={() => setDropdownVisible(!dropdownVisible)}
      ref={AccountSelectorRef}
    >
      {dropdownVisible && (
        <AccountSelectorList
          selectedAccountId={selectedAccount?.id}
          onAccountSelect={handleOnAccountSelect}
        />
      )} 
      <div className="account-selector__image" data-for="ttip-solid-top" data-tip={selectedAccount?.currency?.name}>
        {selectedAccount && (
          <h1 className="account-selector__image-symbol">{selectedAccount.currency?.symbol}</h1>
        )}
      </div>
      {selectedAccount
        ? (
          <div className="account-selector__info">
            <div>
              <h4 className="account-selector__name" data-for="ttip-solid-right" data-tip="Nombre de la cuenta">{selectedAccount.name}</h4>
              <p className="account-selector__date" data-for="ttip-solid-right" data-tip="Fecha de creacion">
                <small>{dayjs(selectedAccount.created_at).format('DD MMM YYYY')}</small>
              </p>
            </div>
            <h4 className="account-selector__funds">
              {selectedAccount.currency?.symbol} {selectedAccount.funds}
            </h4>

          <h6 className="account-selector__id" data-for="ttip-solid-right" data-tip="NÃºmero de cuenta">#{selectedAccount.id}</h6>
          </div>
        )
        : (
          <div className="account-selector__placeholder">
            <h4 className="account-selector__placeholder-text">Selecciona una cuenta</h4>
          </div>
        )}
    </div>
  )
}

export default AccountSelector;