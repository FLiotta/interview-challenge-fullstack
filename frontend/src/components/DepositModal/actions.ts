// @Project
import { Account } from 'interfaces';

export const DEPOSITMODAL_OPEN = '[DEPOSIT MODAL] OPEN';
export const DEPOSITMODAL_CLOSE = '[DEPOSIT MODAL] CLOSE';

export const openDepositModal = (acc: Account) => ({
  type: DEPOSITMODAL_OPEN,
  payload: { acc }
})

export const closeDepositModal = () => ({
  type: DEPOSITMODAL_CLOSE
})