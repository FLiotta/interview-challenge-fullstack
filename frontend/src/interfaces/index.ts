// @Packages
import { ThunkDispatch } from "redux-thunk";

// @Project
import { IStoreState } from 'reducers/rootReducer';
import { AnyAction } from "redux";

export interface IThunkDispatch extends ThunkDispatch<IStoreState, any, AnyAction> {}

export interface BackendResponse<T> {
  error?: string
  message?: string
  data: T | T[] | any
  metadata?: {
    total: number
  }
}

export interface Profile {
  id: number
  username: string,
  first_name?: string
  last_name?: string
  email?: string
  is_superuser: boolean,
  is_staff: boolean
}

export interface Currency {
  id: number,
  name: string,
  symbol: string
}

export interface Account {
  id: number,
  name?: string,
  currency?: Currency
  currency_id?: number,
  funds: number,
  owner_id?: number,
  owner?: any
}

export interface Operation {  
  id: number
  amount: number
  currency: Currency
  operation_type: 'transfer' | 'deposit'
  receiver: any
  sender: any
  date: string | Date
}