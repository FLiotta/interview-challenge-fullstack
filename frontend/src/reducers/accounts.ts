// @Packages
import { AnyAction } from "redux"

// @Project
import { Account } from 'interfaces';
import {
  ACCOUNTS_FETCH,
  ACCOUNTS_UPDATE_PRICE,
  ACCOUNTS_APPEND,
  ACCOUNTS_RESTART
} from 'actions/accounts';

export interface IState { 
  data: Account[],
  total: number
}

const defaultState: IState = {
  data: [],
  total: 0
}

export default (state = defaultState, action: AnyAction) => {
  switch(action.type) {
    case ACCOUNTS_FETCH:
      return action.payload;
    case ACCOUNTS_UPDATE_PRICE:
      const tempState: IState = Object.assign(state);
      const accountIndex: number = tempState.data.findIndex((acc) => acc.id === action.payload.accountId);

      if(action.payload.operation === 'increase') {
        tempState.data[accountIndex].funds += action.payload.amount;
      } else {
        tempState.data[accountIndex].funds -= action.payload.amount;
      }

      return {
        ...state,
        data: tempState.data
      };
    case ACCOUNTS_APPEND:
      return {
        data: [
          ...state.data,
          ...action.payload
        ],
        total: state.total + action.payload.length
      }
    case ACCOUNTS_RESTART:
      return defaultState;
    default:
      return state;
  } 
}