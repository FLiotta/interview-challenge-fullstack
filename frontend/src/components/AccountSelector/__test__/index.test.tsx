// @Packages
import { render, screen } from 'jest-utils';

// @Project
import { appendAccounts } from 'actions/accounts';
import store from 'store';
import AccountSelector from '../index';
import { Account } from 'interfaces';
import { fireEvent, getByTestId } from '@testing-library/react';

const accounts: Account[] = [
  {
    id: 1,
    owner_id: 1,
    name: "Mi billetera",
    currency_id: 2,
    funds: 500,
    deposit_address: "MKWEQROLQ"
  },
  {
    id: 2,
    owner_id: 1,
    name: "Ahorros",
    currency_id: 2,
    funds: 501,
    deposit_address: "MQKWEMQWX"
  }
];

describe('AccountSelector component', () => {
  beforeAll(() => {
    store.dispatch(appendAccounts(accounts));
  })
  
  it('Should match snapshot', () => {
    const { container } = render(
      <AccountSelector
        onAccountSelect={jest.fn}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it('Opens selector on placeholder click', () => {
    render(
      <AccountSelector
        onAccountSelect={jest.fn}
      />
    );

    const selector = screen.getByTestId('account-selector-testid');

    fireEvent.click(selector);

    const dropdown = getByTestId(selector, 'account-dropdown-testid');

    expect(dropdown).toBeInTheDocument();
  });

  it('Loads accounts from redux store', () => {
    const mockFn = jest.fn();
    
    render(
      <AccountSelector
        onAccountSelect={mockFn}
      />
    );

    const selector = screen.getByTestId('account-selector-testid');

    fireEvent.click(selector);

    const accountsItems = screen.queryAllByTestId("account-item-testid");

    expect(accountsItems.length).toBe(accounts.length)
  });

  it('fires onAccountSelect', () => {
    const mockFn = jest.fn();
    
    render(
      <AccountSelector
        onAccountSelect={mockFn}
      />
    );

    const selector = screen.getByTestId('account-selector-testid');

    fireEvent.click(selector);

    const accountsItems = screen.queryAllByTestId("account-item-testid");

    fireEvent.click(accountsItems[0]);

    expect(mockFn).toHaveBeenCalledTimes(1);
  });
})