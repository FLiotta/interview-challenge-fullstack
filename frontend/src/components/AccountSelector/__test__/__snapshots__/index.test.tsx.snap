// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountSelector component Should match snapshot 1`] = `
<div>
  <div
    class="account-selector"
    data-testid="account-selector-testid"
  >
    <div
      class="account-selector__image"
      data-for="ttip-solid-top"
    />
    <div
      class="account-selector__placeholder"
    >
      <h4
        class="account-selector__placeholder-text"
      >
        Selecciona una cuenta
      </h4>
    </div>
  </div>
</div>
`;
